url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/org/jgraph/plaf/basic/BasicGraphUI.java

```diff
@@ -0,0 +1,3220 @@
+						isContextVisible = contextViews != null
+								&& contextViews.length < MAXCELLS
+								&& (!event.isControlDown() || !graph
+										.isCloneable());
+						event.consume();
+					}
+				}
+				// Analyze for common parent
+				if (graph.isMoveIntoGroups() || graph.isMoveOutOfGroups()) {
+					Object[] cells = context.getCells();
+					Object ignoreGroup = graph.getModel().getParent(cells[0]);
+					for (int i = 1; i < cells.length; i++) {
+						Object tmp = graph.getModel().getParent(cells[i]);
+						if (ignoreGroup != tmp) {
+							ignoreGroup = null;
+							break;
+						}
+					}
+					if (ignoreGroup != null)
+						ignoreTargetGroup = graph.getGraphLayoutCache()
+								.getMapping(ignoreGroup, false);
+				}
+			}
+		}
+
+		/**
+		 * Hook for subclassers to return a different view for a mouse click at
+		 * <code>pt</code>. For example, this can be used to return a leaf
+		 * cell instead of a group.
+		 */
+		protected CellView findViewForPoint(Point2D pt) {
+			double snap = graph.getTolerance();
+			Rectangle2D r = new Rectangle2D.Double(pt.getX() - snap, pt.getY()
+					- snap, 2 * snap, 2 * snap);
+			for (int i = 0; i < views.length; i++)
+				if (views[i].intersects(graph, r))
+					return views[i];
+			return null;
+		}
+
+		/**
+		 * Used for move into group to find the target group.
+		 */
+		protected CellView findUnselectedInnermostGroup(double x, double y) {
+			Object[] cells = graph.getDescendants(graph.getRoots());
+			for (int i = cells.length - 1; i >= 0; i--) {
+				CellView view = graph.getGraphLayoutCache().getMapping(
+						cells[i], false);
+				if (view != null && !view.isLeaf()
+						&& !context.contains(view.getCell())
+						&& view.getBounds().contains(x, y))
+					return view;
+			}
+			return null;
+		}
+
+		protected void startDragging(MouseEvent event) {
+			isDragging = true;
+			if (graph.isDragEnabled()) {
+				int action = (event.isControlDown() && graph.isCloneable()) ? TransferHandler.COPY
+						: TransferHandler.MOVE;
+				TransferHandler th = graph.getTransferHandler();
+				setInsertionLocation(event.getPoint());
+				try {
+					th.exportAsDrag(graph, event, action);
+				} catch (Exception ex) {
+					// Ignore
+				}
+			}
+		}
+
+		/**
+		 * @return Returns the parent graph scrollpane for the specified graph.
+		 */
+		public Component getFirstOpaqueParent(Component component) {
+			if (component != null) {
+				Component parent = component;
+				while (parent != null) {
+					if (parent.isOpaque() && !(parent instanceof JViewport))
+						return parent;
+					parent = parent.getParent();
+				}
+			}
+			return component;
+		}
+
+		protected void initOffscreen() {
+			if (!graph.isXorEnabled()) {
+				return;
+			}
+			try {
+				offgraphics = graph.getOffgraphics();
+			} catch (Exception e) {
+				offgraphics = null;
+			} catch (Error e) {
+				offgraphics = null;
+			}
+		}
+
+		/** Process mouse dragged event. */
+		public void mouseDragged(MouseEvent event) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 70. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.